### PR TITLE
Git annex tests

### LINKED ISSUE #11: Test git-annex
#1 needs tests that can be run in CI, especially as it grows in complexity.

### PR DESCRIPTION
**Pull request by kousu**
Fixes #11

### CODE REVIEW

File: integrations/git_annex_test.go
Comment (kousu) [inline integrations/git_annex_test.go:177]: Actually `chmod -R +w` is rather

```suggestion
			err = os.Chmod(path, info.Mode()|0200)
```

File: integrations/git_annex_test.go
Comment (kousu) [inline integrations/git_annex_test.go:201]:
```suggestion
	f.WriteString("*  annex.largefiles=nothing")
	f.WriteString("*.bin  filter=annex annex.largefiles=anything")
```

File: integrations/git_annex_test.go
Comment (kousu) [inline integrations/git_annex_test.go:208]: inline this

**Comment by kousu** [comment]
I've finally expanded on this to a useful point. As compared to the last time I asked for a review (ec094e9ffd28af17bf0c5691c4552db3e08e4d51) from you @mguaypaq, this now (6d60a03b57875b68ee970d50b7b3fc3d8494707b) tests different combinations of permissions (and one (1) protocol):

* {private,public} x {owner, writer, reader} x {ssh}

You can see it in action [here](https://github.com/neuropoly/gitea/actions/runs/2835531817):

Screenshot 2022-08-10 at 16-34-50 Test Git Annex · neuropoly_gitea@eb7391a


There are some things I am unhappy with, but I think I'm going to merge it as-is, because I want to build on this so I can feel confident finally applying your feedback in #6. I would appreciate your eyes taking a look though, even after it's merged.


In particular, I'm super unhappy with how much copy-pasting I ended up doing. For example these are duplicates:

<details><summary>Deploying the "fixture"</summary>

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L77-L111

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L339-L373

</details>

as are each of the sub-cases:

<details><summary><code>git annex init</code> tests</summary>

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L455-L459

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L505-L509

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L243-L247

</details>

The reason it's unrolled like this is because in _some_ of the cases, we expect an error instead:

<details><summary><code>git annex init</code> test</summary>

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L555-L559

</details>

There's also these two sections which are almost-but-not-quite identical; if we expand these tests to cover testing Orgs, those'll be even more of the same:

<details><summary>Repo creation</summary>

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L67-L75

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L330-L337

</details>

I tried really hard to break this up into subroutines, but I couldn't see how to do it. The trouble seems to be is that each subcase, each `t.Run("something", func(*testing.T) { ... `, takes a new `func` which doesn't return anywhere

Also there's this awkward part where I sneakily swap out the user without swapping out the repo, because I wanted to test that `git annex copy` (which uses git-annex-shell + rsync) would work right, which I couldn't do if it was just going to get blocked at `git clone`. A second idea about how to handle this would be most appreciated:

<details><summary>Juggling <code>withAnnexCtxKeyFile</code></summary>

https://github.com/neuropoly/gitea/blob/9a3abffbe1e40d6d42ab6bb92cbb801ee70b69c0/integrations/git_annex_test.go#L278-L297

</details>

File: integrations/git_annex_test.go
Comment (mguaypaq) [inline integrations/git_annex_test.go:814]: A couple of problems with this code:
- A nit: the buffers `b1` and `b2` are created inside the loop, so new ones are made at each iteration. It would be more typical (I think?) to create them outside the loop so they can be reused.
- A more serious one: `f1.Read(b1)` and `f2.Read(b2)` could theoretically make partial reads even before EOF is reached, so they could get out of sync.
- It's not a problem here, but I don't like the full comparison of `b1` and `b2`, rather than a slice of each corresponding to the part that's actually been filled. (If the buffers get re-created at each iteration, you compare a bunch of zero-initialized bytes; if the buffers get re-used, you re-compare a bunch of nonzero bytes that you already compared in a previous iteration.)

My suggestion (though I haven't compiled it, let alone tested it):
```suggestion
	b1 := make([]byte, chunkSize)
	b2 := make([]byte, chunkSize)

	for {
		// Read as much as possible from f1, up to chunkSize
		n1 := 0
		for n1 < chunkSize {
			n, err := f1.Read(b1[n1:])
			n1 += n
			if err == io.EOF {
				break
			}
			if err != nil {
				return false, err
			}
		}

		// Read as much as possible from f2, up to chunkSize
		n2 := 0
		for n2 < chunkSize {
			n, err := f2.Read(b2[n2:])
			n2 += n
			if err == io.EOF {
				break
			}
			if err != nil {
				return false, err
			}
		}

		if n1 == 0 && n2 == 0 {
			// Both files are EOF, all bytes have been compared
			return true, nil
		}

		// If EOF is reached, the next iteration will have n1 == 0 && n2 == 0
		if !bytes.Equal(b1[:n1], b2[:n2]) {
			return false, nil
		}
	}
```

File: integrations/git_annex_test.go
Comment (mguaypaq) [inline integrations/git_annex_test.go:833]: I'd prefer to `require.NoError` for these writes as well? No need to worry about partial writes, because those will have a non-nil error.

File: integrations/git_annex_test.go
Comment (mguaypaq) [inline integrations/git_annex_test.go:819]: Nit: this doesn't actually add a non-annexed file (which is fine, but the docs should match the code).

File: integrations/git_annex_test.go
Comment (mguaypaq) [inline integrations/git_annex_test.go:868]: I like that you're being careful!

File: integrations/git_annex_test.go
Comment (mguaypaq) [inline integrations/git_annex_test.go:889]: Nit: since this is the only use of the `bare` variable, why not put `keyformat = "hashdirlower"` instead of `bare = true` above, and same for `bare = false`?

File: integrations/git_annex_test.go
Comment (mguaypaq) [inline integrations/git_annex_test.go:929]: This should happen _after_ saving the previous value, right?

**Review by mguaypaq — APPROVED** [review]
Thanks for this!

I know there's a lot of copy-pasting, but this is testing code, where I think "explicit is better than implicit" is even more true. As a reader, especially if I was faced with a failing test, I would really appreciate having so few layers of abstraction!

So, I do like how you've structured the tests. The helper functions you defined are a good level of abstraction. And in a sense, you're not very far from having an array of tests defined by test protocols, if you squint a tiny bit.

Re: the multi-user scenario and being sneaky, I don't have any ideas that spring to mind, sorry. But even a sneaky test is better than no test (and can be removed if it ends up causing headaches).

Most of my comments are just comments or nitpicks, but for `filecmp` I think I have a more correct suggestion (and I'm puzzled as to why this isn't in the standard library, but I didn't find it either).

**Comment by kousu** [comment]
Thank you @mguaypaq, this review is excellent. I'll integrate your ideas today and then merge where it gets to.

I'm glad at least your second opinion thinks that the shortcuts aren't that bad. They really bugged me but if you're happy I can be happy.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:889]: In this one instance, I think that would be a lot less clear. The relationship is due to obscure legacy compatibility concerns: https://git-annex.branchable.com/internals/hashing/. I'm hoping that someone reading this, then that link, then this again will be able to tell why this has to be here and I think inlining it would obscure the reasoning. It would for me at least.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:868]: Thanks!

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:819]: It adds `.gitattributes`! Which isn't much of a file, but it's something.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:833]: Yes you're right!

Actually I think I'm going to rewrite this to return an `error` instead of passing in `t`. I started doing that halfway and only using `require.whatever(t, ...)` inside the `t.Run()`s, but this was one of the earlier subroutines I wrote. That was the way the Gitea authors wrote `git_tests.go`, but it's a lot more flexible if the helpers don't think about the tests.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:54]:
```suggestion
```

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:70]: Alright, good call.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:62]: Ah, no, I don't. I thought I did but I was confused. If I drop this, and add a print to RemoveAll, you can see the test harness uses `util.RemoveAll()` to clean up previous tests before it starts:

<details><summary>patch</summary>

```patch
diff --git a/integrations/git_annex_test.go b/integrations/git_annex_test.go
index 313cb2875..40983ca44 100644
--- a/integrations/git_annex_test.go
+++ b/integrations/git_annex_test.go
@@ -23,12 +23,10 @@ import (
        "errors"
        "fmt"
        "io"
-       "io/fs"
        "math/rand"
        "net/url"
        "os"
        "path"
-       "path/filepath"
        "regexp"
        "strings"
 
@@ -59,7 +57,6 @@ func TestGitAnnex(t *testing.T) {
        // repo, you need to edit its .Reponame or just ignore it and write "username/reponame.git"
 
        onGiteaRun(t, func(t *testing.T, u *url.URL) {
-               defer annexUnlockdown() // workaround https://git-annex.branchable.com/internals/lockdown/
 
                t.Run("Public", func(t *testing.T) {
                        defer PrintCurrentTest(t)()
@@ -898,32 +895,6 @@ func annexObjectPath(repoPath string, file string) (string, error) {
        return path.Join(repoPath, "annex", "objects", keyHashPrefix, annexKey, annexKey), nil
 }
 
-/*
-Do chmod -R +w $REPOS in order to handle https://git-annex.branchable.com/internals/lockdown/:
-
-> (The only bad consequence of this is that rm -rf .git doesn't work unless you first run chmod -R +w .git)
-
-Without, these tests can only be run once, because they reuse `gitea-repositories/`
-folder and will balk at finding pre-existing partial repos.
-*/
-func annexUnlockdown() {
-       filepath.WalkDir(setting.RepoRootPath, func(path string, d fs.DirEntry, err error) error {
-               if err == nil {
-                       // 0200 == u+w, in octal unix permission notation
-                       info, err := d.Info()
-                       if err != nil {
-                               return err
-                       }
-
-                       err = os.Chmod(path, info.Mode()|0200)
-                       if err != nil {
-                               return err
-                       }
-               }
-               return nil
-       })
-}
-
 /* like withKeyFile(), but automatically sets it the account given in ctx for use by git-annex */
 func withAnnexCtxKeyFile(t *testing.T, ctx APITestContext, callback func()) {
        os.Setenv("GIT_ANNEX_USE_GIT_SSH", "1") // withKeyFile works by setting GIT_SSH_COMMAND, but git-annex only respects that if this is set
diff --git a/modules/util/remove.go b/modules/util/remove.go
index 95024e11f..bfb9a9782 100644
--- a/modules/util/remove.go
+++ b/modules/util/remove.go
@@ -6,6 +6,7 @@ package util
 
 import (
        "os"
+       "fmt"
        "runtime"
        "syscall"
        "time"
@@ -46,6 +47,8 @@ func Remove(name string) error {
 
 // RemoveAll removes the named file or directory with at most 5 attempts.
 func RemoveAll(name string) error {
+
+       fmt.Printf("RemoveAll(%s)\n", name)
        var err error
 
        for i := 0; i < 5; i++ {
```

</details>

<details><summary>test</summary>

```
p115628@joplin:~/src/neurogitea/gitea$ make test-sqlite#TestGitAnnex.Public.Reader.SSH.Init
RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/tmp/uploads)
RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/data/integrations/gitea-integration-sqlite/tmp/local-repo)
=== TestGitAnnex (integrations/git_helper_for_declarative_test.go:117)
RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/gitea-repositories)
=== TestGitAnnex/Public (integrations/git_annex_test.go:62)
RemoveAll(/tmp/gitea-annex-public4143071661)
RemoveAll(/tmp/index3724207135)
RemoveAll(/tmp/index2224850679)
RemoveAll(/tmp/annex-public2371856901)
RemoveAll(/tmp/key-file88687326)
=== TestGitAnnex/Public/Reader (integrations/git_annex_test.go:212)
=== TestGitAnnex/Public/Reader/SSH (integrations/git_annex_test.go:215)
RemoveAll(/tmp/key-file3426719213)
=== TestGitAnnex/Public/Reader/SSH/Init (integrations/git_annex_test.go:241)
RemoveAll(/tmp/key-file594703350)
RemoveAll(/tmp/annex-public253285415)
+++ TestGitAnnex/Public is a slow test (took 20.893163374s)
+++ TestGitAnnex is a slow test (took 20.97350748s)
PASS
```

</details>

the key lines being

> === TestGitAnnex (integrations/git_helper_for_declarative_test.go:117)
> RemoveAll(/home/GRAMES.POLYMTL.CA/p115628/src/neurogitea/gitea/integrations/gitea-integration-sqlite/gitea-repositories)

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:929]: whoops yep :)

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:321]: I was. But that doesn't have to go into the committed patch. I can just isolate it with `git add -p`

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:634]: Yeah I mean, this one is sketchier. The output looks like, say,

```
whereis derivatives/labels/sub-beijingVerio04/anat/sub-beijingVerio04_T2w_RPI_r_seg-manual.nii.gz (5 copies) 
  	5a5447a8-a9b8-49bc-8276-01a62632b502 -- [amazon]
   	5cdba4fc-8d50-4e89-bb0c-a3a4f9449666 -- julien@julien-macbook.local:~/code/spine-generic/data-multi-subject
   	fc75435d-eb11-4c5a-9b68-debf6e68df2a -- [origin]
   	9e4d13f3-30e1-4a29-8b86-670879928606 -- alex@NeuroPoly-MacBook-Pro.local:~/data/data-multi-subject
   	e405e14e-33b2-4a35-b7a7-3eeec054f0d4 -- sebeda@joplin.neuro.polymtl.ca:/mnt/nvme/sebeda/data-multi-subject
```

I suppose I could add `\n` at the ends, but then what if that breaks down in some cases? Leaving it unattached is easier...

I'll add a comment.

Comment (kousu) [reply integrations/git_annex_test.go:634]: I'm hoping `[origin]\n` should be enough to anchor it

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:749]: I don't deserve the credit. I copied this from 

https://github.com/neuropoly/gitea/blob/37cf5c9e9b9f18b5d56df5fa9db52d932659e3b3/integrations/git_test.go#L326-L332

I just took out the git-specific parts.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:766]: Whoops I did. I guess I was thinking `2**12`.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:745]: I guess so! I didn't even read this code.

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:757]: I think you're right but I'm going to leave it because it was like that when I found it!

File: integrations/git_annex_test.go
Comment (kousu) [reply integrations/git_annex_test.go:803]: I got the outline off this code off StackOverflow, which went through a few revisions itself. Someone else had the [same thought](https://stackoverflow.com/questions/29505089/how-can-i-compare-two-files-in-golang/30038571#comment113030456_30038571) as you.

Hm.

Can we replace your inner loops by [io.ReadFull](https://pkg.go.dev/io#ReadFull)?

> On return, n == len(buf) if and only if err == nil

except how does it handle EOFs?

I'll play with it.

Comment (kousu) [reply integrations/git_annex_test.go]: Here's what I've got?

https://github.com/neuropoly/gitea/blob/a3c5ab7b3636e2723204ad46f835c6304db1b099/modules/util/filecmp.go#L9-L79

Comment (mguaypaq) [reply integrations/git_annex_test.go]: I like it!

Typo on line 25: `file1` should be `file2`.

Also, I think it would be typical go style to document that `err != nil` implies `same == false`, which is true of your implementation, rather than the disclaimer you wrote at the top.

Comment (kousu) [reply integrations/git_annex_test.go]: Oh no! Drat. That's what I get for adding `os.SameFile` after doing most of my testing late at night.

Thank you for catching that.

**Pull request merged**
